### PR TITLE
Fix Issue 17153 – Make std.container.array @nogc

### PR DESCRIPTION
**Pull request by JackStouffer**
Despite `Array` not using the GC for allocating its elements, it was still unusable for `@nogc` code because of its use of `throw` and `enforce`.

This PR turns those into `asserts` and makes the majority of the functions `@nogc`.

### CODE REVIEW

**Comment by JackStouffer** [comment]
@andralex requesting your review because this is a breaking change.

File: std/container/array.d [hunk 1]
```diff
@@ -142,26 +142,26 @@ private struct RangeT(A)
 
         E moveFront()
         {
-            version (assert) if (empty || _a >= _outer.length) throw new RangeError();
+            assert(!empty || _a < _outer.length);
```
Comment (aG0aep6G) [inline std/container/array.d]: Unless I'm having a dumb moment, this should use `&&`. `!(foo || bar)` = `!foo && !bar`. Same mistake throughout.

**Comment by JackStouffer** [comment]
@wilzbach The coverage report that DMD generates and the coverage that appears on codecov are different.

**Comment by wilzbach** [comment]
> @wilzbach The coverage report that DMD generates and the coverage that appears on codecov are different.

That looks pretty weird and I am not sure how to fix this. Maybe @stevepeak (from CodeCov) knows the reason for this issue with CodeCov?

**Comment by JackStouffer** [comment]
> but I am not sure how far @andralex is on his journey to make it possible to use exceptions in @nogc?

Requires `@safe` RC which requires escape analysis which requires DIP1000 to be completed.

File: std/container/array.d [hunk 2]
```diff
@@ -355,8 +354,11 @@ if (!is(Unqual!T == bool))
                  * than realloc.
                  */
                 immutable oldLength = length;
-                auto newPayload =
-                    enforce(cast(T*) malloc(sz))[0 .. oldLength];
+
+                auto newPayloadPtr = cast(T*) malloc(sz);
```
Comment (ZombineDev) [inline std/container/array.d]: Consider using `pureMalloc` next time ;)

File: std/container/array.d [hunk 3]
```diff
@@ -1471,49 +1489,49 @@ if (is(Unqual!T == bool))
...
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to set the back of an empty Array");
             _outer[_b - 1] = value;
         }
         /// Ditto
         T moveBack()
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to mpve the back of an empty Array");
```
Comment (ZombineDev) [inline std/container/array.d]: move typo

File: std/container/array.d [hunk 4]
```diff
@@ -355,8 +354,11 @@ if (!is(Unqual!T == bool))
                  * than realloc.
                  */
                 immutable oldLength = length;
-                auto newPayload =
-                    enforce(cast(T*) malloc(sz))[0 .. oldLength];
+
+                auto newPayloadPtr = cast(T*) malloc(sz);
+                assert(newPayloadPtr, "std.container.Array.reserve failed to allocate memory");
```
Comment (andralex) [inline std/container/array.d]: Let's do this: ` newPayloadPtr || assert(false, "std.container.Array.reserve failed to allocate memory");`

File: std/container/array.d [hunk 5]
```diff
@@ -374,8 +376,9 @@ if (!is(Unqual!T == bool))
                 /* These can't have pointers, so no need to zero
                  * unused region
                  */
-                auto newPayload =
-                    enforce(cast(T*) realloc(_payload.ptr, sz))[0 .. length];
+                auto newPayloadPtr = cast(T*) realloc(_payload.ptr, sz);
+                assert(newPayloadPtr, "std.container.Array.reserve failed to allocate memory");
+                auto newPayload = newPayloadPtr[0 .. length];
```
Comment (andralex) [inline std/container/array.d]: Same here, end the app if no memory.

File: std/container/array.d [hunk 6]
```diff
@@ -547,7 +550,8 @@ Complexity: $(BIGOH 1)
             bool overflow;
             const sz = mulu(elements, T.sizeof, overflow);
             if (overflow) assert(0);
-            auto p = enforce(malloc(sz));
+            auto p = malloc(sz);
+            assert(p);
```
Comment (andralex) [inline std/container/array.d]: Same here, end the app if no memory.

File: std/container/array.d [hunk 7]
```diff
@@ -805,7 +809,7 @@ Complexity: $(BIGOH log(n)).
      */
     void removeBack()
     {
-        enforce(!empty);
+        assert(!empty);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

**Comment by stevepeak** [comment]
Hey @wilzbach can you check if the [webhook is setup](https://github.com/dlang/phobos/settings/hooks) and sending to Codecov? Status was waiting on a webhook. Thank you!

File: std/container/array.d [hunk 8]
```diff
@@ -855,7 +859,7 @@ Complexity: $(BIGOH n + m), where $(D m) is the length of $(D stuff)
     if (isImplicitlyConvertible!(Stuff, T))
     {
         import std.conv : emplace;
-        enforce(r._outer._data is _data && r._a <= length);
+        assert(r._outer._data is _data && r._a <= length);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 9]
```diff
@@ -872,7 +876,7 @@ Complexity: $(BIGOH n + m), where $(D m) is the length of $(D stuff)
     if (isInputRange!Stuff && isImplicitlyConvertible!(ElementType!Stuff, T))
     {
         import std.conv : emplace;
-        enforce(r._outer._data is _data && r._a <= length);
+        assert(r._outer._data is _data && r._a <= length);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 10]
```diff
@@ -897,9 +901,9 @@ Complexity: $(BIGOH n + m), where $(D m) is the length of $(D stuff)
         else
         {
             import std.algorithm.mutation : bringToFront;
-            enforce(_data);
+            assert(_data);
```
Comment (andralex) [inline std/container/array.d]: When could this fail?

File: std/container/array.d [hunk 11]
```diff
@@ -897,9 +901,9 @@ Complexity: $(BIGOH n + m), where $(D m) is the length of $(D stuff)
...
             immutable offset = r._a;
-            enforce(offset <= length);
```
Comment (andralex) [inline std/container/array.d]: When could this fail?

File: std/container/array.d [hunk 12]
```diff
@@ -911,10 +915,10 @@ Complexity: $(BIGOH n + m), where $(D m) is the length of $(D stuff)
     size_t insertAfter(Stuff)(Range r, Stuff stuff)
     {
         import std.algorithm.mutation : bringToFront;
-        enforce(r._outer._data is _data);
+        assert(r._outer._data is _data);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 13]
```diff
@@ -911,10 +915,10 @@ Complexity: $(BIGOH n + m), where $(D m) is the length of $(D stuff)
...
         // TODO: optimize
         immutable offset = r._b;
-        enforce(offset <= length);
+        assert(offset <= length);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 14]
```diff
@@ -925,7 +929,7 @@ Complexity: $(BIGOH n + m), where $(D m) is the length of $(D stuff)
     size_t replace(Stuff)(Range r, Stuff stuff)
     if (isInputRange!Stuff && isImplicitlyConvertible!(ElementType!Stuff, T))
     {
-        enforce(r._outer._data is _data);
+        assert(r._outer._data is _data);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 15]
```diff
@@ -947,7 +951,7 @@ Complexity: $(BIGOH n + m), where $(D m) is the length of $(D stuff)
     size_t replace(Stuff)(Range r, Stuff stuff)
     if (isImplicitlyConvertible!(Stuff, T))
     {
-        enforce(r._outer._data is _data);
+        assert(r._outer._data is _data);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 16]
```diff
@@ -977,9 +981,9 @@ $(D r)
     {
         import std.algorithm.mutation : copy;
 
-        enforce(r._outer._data is _data);
-        enforce(_data.refCountedStore.isInitialized);
-        enforce(r._a <= r._b && r._b <= length);
+        assert(r._outer._data is _data);
+        assert(_data.refCountedStore.isInitialized);
+        assert(r._a <= r._b && r._b <= length);
```
Comment (andralex) [inline std/container/array.d]: These three do not belong with this PR, they're a (breaking) change in philosophy.

File: std/container/array.d [hunk 17]
```diff
@@ -1471,49 +1489,49 @@ if (is(Unqual!T == bool))
         /// Ditto
         @property T front()
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to access the front of an empty Array");
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 18]
```diff
@@ -1471,49 +1489,49 @@ if (is(Unqual!T == bool))
...
             return _outer[_a];
         }
         /// Ditto
         @property void front(bool value)
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to set the front of an empty Array");
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 19]
```diff
@@ -1471,49 +1489,49 @@ if (is(Unqual!T == bool))
...
             _outer[_a] = value;
         }
         /// Ditto
         T moveFront()
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to move the front of an empty Array");
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 20]
```diff
@@ -1471,49 +1489,49 @@ if (is(Unqual!T == bool))
...
             return _outer.moveAt(_a);
         }
         /// Ditto
         void popFront()
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to popFront an empty Array");
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 21]
```diff
@@ -1471,49 +1489,49 @@ if (is(Unqual!T == bool))
...
             ++_a;
         }
         /// Ditto
         @property T back()
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to access the back of an empty Array");
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 22]
```diff
@@ -1471,49 +1489,49 @@ if (is(Unqual!T == bool))
...
             return _outer[_b - 1];
         }
         /// Ditto
         @property void back(bool value)
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to set the back of an empty Array");
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d (hunk 3 above)
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 23]
```diff
@@ -1471,49 +1489,49 @@ if (is(Unqual!T == bool))
...
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to mpve the back of an empty Array");
             return _outer.moveAt(_b - 1);
         }
         /// Ditto
         void popBack()
         {
-            enforce(!empty);
+            assert(!empty, "Attempting to popBack an empty Array");
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 24]
```diff
@@ -1689,7 +1710,7 @@ if (is(Unqual!T == bool))
      */
     Range opSlice(size_t a, size_t b)
     {
-        enforce(a <= b && b <= length);
+        assert(a <= b && b <= length);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 25]
```diff
@@ -1708,14 +1729,14 @@ if (is(Unqual!T == bool))
      */
     @property bool front()
     {
-        enforce(!empty);
+        assert(!empty);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 26]
```diff
@@ -1708,14 +1729,14 @@ if (is(Unqual!T == bool))
...
         return data.ptr[0] & 1;
     }
 
     /// Ditto
     @property void front(bool value)
     {
-        enforce(!empty);
+        assert(!empty);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 27]
```diff
@@ -1732,14 +1753,14 @@ if (is(Unqual!T == bool))
     /// Ditto
     @property bool back()
     {
-        enforce(!empty);
+        assert(!empty);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 28]
```diff
@@ -1732,14 +1753,14 @@ if (is(Unqual!T == bool))
...
         return cast(bool)(data.back & (cast(size_t)1 << ((_store._length - 1) % bitsPerWord)));
     }
 
     /// Ditto
     @property void back(bool value)
     {
-        enforce(!empty);
+        assert(!empty);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 29]
```diff
@@ -1767,15 +1788,15 @@ if (is(Unqual!T == bool))
     {
         auto div = cast(size_t) (i / bitsPerWord);
         auto rem = i % bitsPerWord;
-        enforce(div < data.length);
+        assert(div < data.length);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 30]
```diff
@@ -1767,15 +1788,15 @@ if (is(Unqual!T == bool))
...
         return cast(bool)(data.ptr[div] & (cast(size_t)1 << rem));
     }
     /// ditto
     void opIndexAssign(bool value, size_t i)
     {
         auto div = cast(size_t) (i / bitsPerWord);
         auto rem = i % bitsPerWord;
-        enforce(div < data.length);
+        assert(div < data.length);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 31]
```diff
@@ -1784,7 +1805,7 @@ if (is(Unqual!T == bool))
...
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

File: std/container/array.d [hunk 32]
```diff
@@ -2050,7 +2071,7 @@ if (is(Unqual!T == bool))
      */
     void removeBack()
     {
-        enforce(_store._length);
+        assert(_store._length);
```
Comment (andralex) [inline std/container/array.d]: Does not belong with this PR, it's a (breaking) change in philosophy.

**Comment by wilzbach** [comment]
> @wilzbach The coverage report that DMD generates and the coverage that appears on codecov are different.

The problem is that we [merge the upstream branch into the PR](https://github.com/dlang/phobos/blob/master/circleci.sh#L68) and thus generate slightly different `.lst` files. This is also the reason why `codecov/project` appears to be fluctuating so often.
I opened a support issue at CodeCov: https://github.com/codecov/support/issues/360

**Comment by JackStouffer** [comment]
@andralex Removed controversial changes

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**